README.md
# DayCY
[![npm](https://img.shields.io/npm/v/daycy.svg)](https://www.npmjs.com/package/daycy/)

Simple date (range) picker for React based on Semantic-UI.

[**Demo / overview of all components**](https://CodeYellowBV.github.io/daycy/)

## Install
```
npm install daycy
```

## Usage
First we will have to include the CSS. The easiest way to do this is to have
your babel setup in such a way that you can import CSS. DayCY also needs the
CSS from Semantic UI.
```
npm install semantic-ui-css
```
```jsx
import 'semantic-ui-css/semantic.min.css';
import 'daycy/dist/daycy.css';
```
This has the added benefit of automatically keeping the CSS in sync with the
version of the package you have installed.

## Internationalization
By default the calendar popup will use English names for the months and days of
the week. To override this you can change the translation function.
```jsx
import { configureTranslation } from 'daycy';

function myTranslationFunction(key) {
    // Insert your logic here
}

configureTranslation(myTranslationFunction);
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.5.2] - 2023-03-23
### Changed
- Upgraded dependencies to improve security

## [0.5.1] - 2021-11-03
### Fixed
- Fixed `MonthPicker` to correctly forward the value to `MonthInput`.
### Changed
- Updated the storybook to properly maintain state for a better demo experience.

## [0.5.0] - 2021-11-03
### Added
- Added components `MonthPicker`, `MonthInput`, and `MonthCalendar` that allow for month selection.

## [0.4.11] - 2021-04-30
### Changed
- Changed all components to leave the part of the date that they do not change the same. So a time picker does not change the date and a date picker does not change the time.

## [0.4.10] - 2020-09-10
### Added
- Added support for `noPopup` prop to all picker components.

## [0.4.9] - 2020-08-28
### Added
- Added support for `disabled` to `DateRangePicker` and `TimeRangePicker`.

## [0.4.8] - 2020-07-30
### Changed
- Changed open/close of calendar/clock to be more tightly coupled to the focus state of the input.

## [0.4.7] - 2020-07-14
### Fixed
- Fix issue with onChange in `DateInput`.

## [0.4.6] - 2020-07-14
### Added
- Added support for more formats for smart typing in `DateInput`.

## [0.4.5] - 2020-07-14
### Fixed
- Fixed some issues with date format pipes in `DateInput`.

## [0.4.4] - 2020-07-14
### Added
- Make `DateInput` support smart typing validation for certain date formats.

## [0.4.3] - 2020-01-20
### Changed
- Make `TimeInput` inherit from `DateInput`.

## [0.4.2] - 2020-01-09
### Fixed
- Fixed issue in clock where AM/PM would not work correctly.

## [0.4.1] - 2020-01-09
### Fixed
- Fixed issue in clock where minutes would display as 60 instead of 0.

## [0.4.0] - 2020-01-09
### Added
- Added `TimeInput`, `TimePicker` and `TimeRangePicker`.

## [0.3.1] - 2019-11-28
### Fixed
- Fixed initial state for `typeValue` in `DateInput`

## [0.3.0] - 2019-11-27
### Added
- Add the ability to type to `DatePicker` and `DateRangePicker`.

## [0.2.0] - 2019-07-24
### Added
- Add `WeekPicker` component.
### Changed
- Change usage of `week.number` translation key to receive argument `week`
instead of `weekNumber`.

## [0.1.5] - 2019-07-23
### Added
- Add `onWeekSelect` property to `DatePicker` to determine what should happen
when someone selects a week in the `DatePicker`.

## [0.1.4] - 2019-07-04
### Added
- Add `includeWeeks` property to both `DatePicker` and `DateRangePicker` for
displaying week numbers to the calendar. For `DateRangePicker` this week number
can also be clicked to select this week as your range.

## [0.1.3] - 2019-06-13
### Added
- Add fluid property to `DatePicker` and always enable it when it is inside a
Semantic UI Form.

## [0.1.2] - 2019-06-04
### Fixed
- Fix `configureTranslation` so that it actually does what it says it does.

## [0.1.1] - 2019-06-04
### Added
- Make `DateRangePicker` always fluid when it is inside a Semantic UI Form.

## [0.1.0] - 2019-05-23
### Added
- First version of this project.

[0.4.10]: https://github.com/CodeYellowBV/daycy/compare/v0.4.9...v0.4.10
[0.4.9]: https://github.com/CodeYellowBV/daycy/compare/v0.4.8...v0.4.9
[0.4.8]: https://github.com/CodeYellowBV/daycy/compare/v0.4.7...v0.4.8
[0.4.7]: https://github.com/CodeYellowBV/daycy/compare/v0.4.6...v0.4.7
[0.4.6]: https://github.com/CodeYellowBV/daycy/compare/v0.4.5...v0.4.6
[0.4.5]: https://github.com/CodeYellowBV/daycy/compare/v0.4.4...v0.4.5
[0.4.4]: https://github.com/CodeYellowBV/daycy/compare/v0.4.3...v0.4.4
[0.4.3]: https://github.com/CodeYellowBV/daycy/compare/v0.4.2...v0.4.3
[0.4.2]: https://github.com/CodeYellowBV/daycy/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/CodeYellowBV/daycy/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/CodeYellowBV/daycy/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com/CodeYellowBV/daycy/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/CodeYellowBV/daycy/compare/0.2.0...v0.3.0
[0.2.0]: https://github.com/CodeYellowBV/daycy/compare/0.1.5...0.2.0
[0.1.5]: https://github.com/CodeYellowBV/daycy/compare/0.1.4...0.1.5
[0.1.4]: https://github.com/CodeYellowBV/daycy/compare/0.1.3...0.1.4
[0.1.3]: https://github.com/CodeYellowBV/daycy/compare/0.1.2...0.1.3
[0.1.2]: https://github.com/CodeYellowBV/daycy/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/CodeYellowBV/daycy/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/CodeYellowBV/daycy/tree/0.1.0


src/Clock.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Popup } from 'semantic-ui-react';
import { DateTime } from 'luxon';

import { translate } from './translate';

const POINTER_LENGTH = {
    hour: '2.5rem',
    minute: '3.5rem',
};
const POINTER_WIDTH = {
    hour: 6,
    minute: 4,
};
const PERIODS = ['am', 'pm'];

function range(start, stop, step = 1) {
    if (stop === undefined) {
        stop = start;
        start = 0;
    }
    const items = [];
    for (let item = start; item < stop; item += step) {
        items.push(item);
    }
    return items;
}

export default class Clock extends Component {
    static propTypes = {
        trigger: PropTypes.node.isRequired,
        open: PropTypes.bool.isRequired,
        value: PropTypes.instanceOf(DateTime),
        onChange: PropTypes.func.isRequired,
        onClose: PropTypes.func,
        translate: PropTypes.func,
    };

    static defaultProps = {
        onClose: () => {},
    };

    state = {
        period: 'am',

        hour: null,
        hourHover: null,
        lastHourHover: 12,

        minuteHover: null,
        lastMinuteHover: 60,
    };

    constructor(...args) {
        super(...args);
        this.renderHour = this.renderHour.bind(this);
        this.renderMinute = this.renderMinute.bind(this);
        this.renderPeriod = this.renderPeriod.bind(this);
    }

    translate(...args) {
        return (this.props.translate || translate)(...args);
    }

    onChange(date) {
        const { value, onChange } = this.props
        if (value) {
            date = date.set({
                year: value.year,
                month: value.month,
                day: value.day,
            })
        }
        onChange(date)
    }

    renderHour(hour) {
        const { hour: selectedHour, period } = this.state;

        const classes = ['button'];

        if (selectedHour !== null) {
            classes.push('small');
        }
        if (hour === selectedHour) {
            classes.push('selected');
        }

        const angle = Math.PI * 2 * (hour - 3) / 12;
        return (
            <div
                key={hour}
                className={classes.join(' ')}
                style={{
                    left: `${6 + 6 * Math.cos(angle)}rem`,
                    top: `${6 + 6 * Math.sin(angle)}rem`,
                    transformOrigin: `${-6 * Math.cos(angle)}rem ${-6 * Math.sin(angle)}rem`,
                }}
                onMouseEnter={() => this.setState({ hourHover: hour })}
                onMouseLeave={() => this.setState({ hourHover: null, lastHourHover: hour })}
                onClick={() => this.setState({ hour })}
            >
                {(period === 'pm' ? 12 : 0) + hour}
            </div>
        );
    }

    renderMinute(minute) {
        const { period, hour } = this.state;

        const classes = ['button'];

        if (hour === null) {
            classes.push('disabled');
        }

        const angle = Math.PI * 2 * (minute - 15) / 60;
        return (
            <div
                key={minute}
                className={classes.join(' ')}
                style={{
                    left: `${6 + 6 * Math.cos(angle)}rem`,
                    top: `${6 + 6 * Math.sin(angle)}rem`,
                    transformOrigin: `${-6 * Math.cos(angle)}rem ${-6 * Math.sin(angle)}rem`,
                }}
                onMouseEnter={() => this.setState({ minuteHover: minute })}
                onMouseLeave={() => this.setState({ minuteHover: null, lastMinuteHover: minute })}
                onClick={() => {
                    this.onChange(DateTime.fromObject({ hour: (hour + (period === 'pm' ? 12 : 0)) % 24, minute }))
                    this.setState({
                        period: 'am',

                        hour: null,
                        hourHover: null,
                        lastHourHover: 12,

                        minuteHover: null,
                        lastMinuteHover: 60,
                    });
                }}
            >
                {minute}
            </div>
        );
    }

    renderPointer({ type, angle, hide = false }) {
        const length = POINTER_LENGTH[type];
        const width = POINTER_WIDTH[type];

        return (
            <div
                className={`pointer ${type} ${hide ? ' hide' : ''}`}
                style={{
                    height: `${width}px`,
                    width: `calc(${length} + ${width}px)`,
                    borderRadius: `${width / 2}px`,
                    left: `calc(6rem - ${width / 2 + 2}px)`,
                    top: `calc(6rem - ${width / 2 + 2}px)`,
                    transformOrigin: `${width / 2}px ${width / 2}px`,
                    transform: `rotate(${angle}rad)`,
                }}
            />
        );
    }

    renderPeriod(period) {
        const { period: selectedPeriod } = this.state;
        return (
            <button
                key={period}
                className={`period ${period}${period === selectedPeriod ? ' selected' : ''}`}
                onClick={() => this.setState({ period })}
            >
                {this.translate(`period.${period}`)}
            </button>
        );
    }

    render() {
        const { trigger, open, onClose } = this.props;
        const {
            period,
            hour, hourHover, lastHourHover,
            minuteHover, lastMinuteHover,
        } = this.state;

        return (
            <Popup
                className="daycy clock"
                flowing
                trigger={trigger}
                open={open}
                on="click"
                onClose={onClose}
                onMouseDown={(e) => e.preventDefault()}
            >
                <div className="face">
                    {range(0, 60, 5).map(this.renderMinute)}
                    {range(1, 13).map(this.renderHour)}
                    <div className={`period-bg ${period}`} />
                    {PERIODS.map(this.renderPeriod)}
                    {this.renderPointer({
                        type: 'minute',
                        angle: Math.PI * 2 * ((minuteHover || lastMinuteHover) - 15) / 60,
                        hide: minuteHover === null,
                    })}
                    {this.renderPointer({
                        type: 'hour',
                        angle: Math.PI * 2 * ((hourHover || hour || lastHourHover) - 3) / 12,
                        hide: (hourHover || hour) === null,
                    })}
                </div>
            </Popup>
        );
    }
}
